src/Misc/MiddleWare.cpp

```diff
@@ -1778,6 +2005,49 @@ static rtosc::Ports middlewareReplyPorts = {
         rEnd},
     {"broadcast:", 0, 0, rBegin; impl.broadcast = true; rEnd},
     {"forward:", 0, 0, rBegin; impl.forward = true; rEnd},
+    {"request-wavetable:sTi:sFi", 0, 0,
+        rBegin;
+        int part, kit, voice;
+
+        string voicePath     = rtosc_argument(msg, 0).s; // it's *not yet* the voice path
+        const bool isFm      = rtosc_argument(msg, 1).T;
+        const int presonance = rtosc_argument(msg, 2).i;
+        bool res = idsFromMsg(voicePath.c_str(), &part, &kit, &voice);
+        assert(res);
+
+        string::size_type pos = voicePath.find("wavetable-rt-params");
+        assert(pos == voicePath.length() - strlen("wavetable-rt-params"));
+        voicePath.resize(pos); // *now* it's the voice location
+
+        OscilGen* oscilGen = static_cast<OscilGen*>(
+                             impl.obj_store.get(voicePath +
+                                                (isFm ? "FMSmp/" : "OscilSmp/")));
+
+        //printf("got: %s -> %s, %s. %s %s %d\n", d.loc, d.message, msg, voicePath.c_str(), isFm ? "true":"false", presonance);
+        assert(oscilGen);
+        assert(!isFm || presonance == 0);
+        const WaveTable* wt = oscilGen->calculateWaveTable(presonance);
+        d.chain((voicePath + "set-wavetable").c_str(), isFm ? "bT" : "bF", sizeof(WaveTable*), &wt);
+
+        impl.waveTableRequestHandler.receivedAdPars(part, kit, voice, isFm);
+        rEnd},
+    {"rt_paste_done:s", 0, 0,
+        rBegin;
+        std::string url = rtosc_argument(msg, 0).s;
+        int part, kit, vc;
+        bool isFm;
+        std::size_t res = idsFromMsg(url.c_str(), &part, &kit, &vc, &isFm);
+        if(res)
+            impl.pasteStateHandler.notifyPasteDone(part, kit, vc, isFm);
+        printf("rPaste done.\n");
+        // if URL ends on "/paste[^/]*", cut before the "/paste"
+        string::size_type last_slash = url.rfind("/paste");
+        assert(last_slash != string::npos);
+        if(!url.compare(last_slash, 6, "/paste"))
+            url.resize(last_slash+1);
+        d.broadcast("/damage", "s", url.c_str());
```

Comment:
With this commit, *every* RT paste (i.e. all except PAD note's non-RT parameter paste) gets its "/damage" delayed until we are here, i.e. until the paste is done (the commit message may not reflect this). However, I think it should be only delayed for any RT paste on non-RT parameters, i.e. only on OscilGen paste. For all other pastes, waiting with "/damage" until the paste has been finished does not make the code more correct, it just increases the time for which the widgets are out-of-date, which is undesired?

Reply:
Right. Only the hacky dance needed for pastes on non-RT data (which involves this pseudo locking) needs a delayed damage message. Delaying the damage message isn't that big of a deal otherwise since the introduced additional latency should be what? a few ms? Hopefully users aren't _that_ demanding :p